bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -298,82 +297,172 @@ bool ReplicaImp::checkSendPrePrepareMsgPrerequisites() {
+                                                            nextRequest->requestSeqNum())) {
+      prePrepareMsg.addRequest(nextRequest->body(), nextRequest->size());
+      clientsManager->addPendingRequest(nextRequest->clientProxyId(), nextRequest->requestSeqNum());
+    }
+    primaryCombinedReqSize -= nextRequest->size();
+  } else if (nextRequest->size() > maxStorageForRequests) {  // The message is too big
+    LOG_ERROR(GL,
+              "Request was dropped because it exceeds maximum allowed size" << KVLOG(
+                  prePrepareMsg.seqNumber(), nextRequest->senderId(), nextRequest->size(), maxStorageForRequests));
+  }
+  delete nextRequest;
+  requestsQueueOfPrimary.pop();
+  return (!requestsQueueOfPrimary.empty() ? requestsQueueOfPrimary.front() : nullptr);
+}
+
+PrePrepareMsg *ReplicaImp::finishAddingRequestsToPrePrepareMsg(PrePrepareMsg *&prePrepareMsg,
+                                                               uint16_t maxSpaceForReqs,
+                                                               uint32_t requiredRequestsSize,
+                                                               uint32_t requiredRequestsNum) {
   if (prePrepareMsg->numberOfRequests() == 0) {
     LOG_INFO(GL, "No client requests added to the PrePrepare batch, delete the message");
     delete prePrepareMsg;
     return nullptr;
   }
-
   prePrepareMsg->finishAddingRequests();
-  LOG_DEBUG(GL, "Consensus batch size" << KVLOG(prePrepareMsg->numberOfRequests()));
+  LOG_DEBUG(GL,
+            KVLOG(prePrepareMsg->seqNumber(),
+                  maxSpaceForReqs,
+                  requiredRequestsSize,
+                  prePrepareMsg->requestsSize(),
+                  requiredRequestsNum,
+                  prePrepareMsg->numberOfRequests()));
   return prePrepareMsg;
 }
 
+PrePrepareMsg *ReplicaImp::buildPrePrepareMessage() {
+  PrePrepareMsg *prePrepareMsg = createPrePrepareMessage();
+  if (!prePrepareMsg) return nullptr;
+
+  uint16_t maxSpaceForReqs = prePrepareMsg->remainingSizeForRequests();
+  ClientRequestMsg *nextRequest = requestsQueueOfPrimary.front();
+  while (nextRequest != nullptr)
+    nextRequest = addRequestToPrePrepareMessage(nextRequest, *prePrepareMsg, maxSpaceForReqs);
+
+  return finishAddingRequestsToPrePrepareMsg(prePrepareMsg, maxSpaceForReqs, 0, 0);
+}
+
+PrePrepareMsg *ReplicaImp::buildPrePrepareMessageByRequestsNum(uint32_t requiredRequestsNum) {
+  PrePrepareMsg *prePrepareMsg = createPrePrepareMessage();
+  if (!prePrepareMsg) return nullptr;
+
+  uint16_t maxSpaceForReqs = prePrepareMsg->remainingSizeForRequests();
+  ClientRequestMsg *nextRequest = requestsQueueOfPrimary.front();
+  while (nextRequest != nullptr && prePrepareMsg->numberOfRequests() < requiredRequestsNum)
+    nextRequest = addRequestToPrePrepareMessage(nextRequest, *prePrepareMsg, maxSpaceForReqs);
+
+  return finishAddingRequestsToPrePrepareMsg(prePrepareMsg, maxSpaceForReqs, 0, requiredRequestsNum);
+}
```

Comment:
Can we solve the code duplication here by using lambdas?
* one for the additional while loop predicate
* one for the `finishAddingRequestsToPrePrepareMsg`